Sources/PackageLoading/PackageBuilder.swift

```diff
@@ -208,15 +208,13 @@ public struct PackageBuilder {
     ///   - includingTestModules: Whether the package's test modules should be loaded.
     public func construct(includingTestModules: Bool) throws -> Package {
         let modules = try constructModules()
-        let testModules: [Module]
-        if includingTestModules {
-            testModules = try constructTestModules(modules: modules)
-        } else {
-            testModules = []
-        }
+        // It is unfortunate that test modules has to be constructed even when they are
+        // not requested by the clients but they might be referenced in the manifest file
+        // so that needs to be validated.
+        let testModules = try constructTestModules(modules: modules)
         try fillDependencies(modules: modules + testModules)
-        let products = try constructProducts(modules, testModules: testModules)
-        return Package(manifest: manifest, path: packagePath, modules: modules, testModules: testModules, products: products)
+        let products = try constructProducts(modules, testModules: includingTestModules ? testModules : [])
+        return Package(manifest: manifest, path: packagePath, modules: modules, testModules: includingTestModules ? testModules : [], products: products)
```

Comment:
Are you sure this won't cause us to start building all tests for all packages by default?